### ISSUE TITLE
Missing method or a bug?

### ISSUE THREAD

**Issue by raminammour** [ChevronETC/DistributedJets.jl#12]
Hello,

Running the examples on version 1.8:

```
A = @blockop DArray(I->[JopDiagonal(rand(2)) for irow in I[1], icol in I[2]], (4,4), workers(), [2,2])
MethodError: no method matching similar(::JetDSpace{Float64, JetBSpace{Float64, JetSpace{Float64, 1}}}, ::Tuple{Int64})
```

Defining these methods **seems** to work but ignores `dims` argument and aliases `R.blkspaces` (I just keyboard monkeyed other JET packages, the alias may be ok?): 
```
@eval DistributedJets begin
Base.similar(R::JetDSpace{T,S}, dims::NTuple{N,Int}) where {T,S,N} = JetDSpace{T,S}(R.blkspaces,R.blkindices,R.indices)
Base.similar(R::JetDSpace, dims::Int...) = similar(R, dims)
end
```

But this aliases `R.blkspaces`, replacing with `similar(R.blkspaces)` leads to:
``` 
On worker 2:
UndefRefError: access to undefined reference
```
(Same happens for `similar(A.jet.dom.blkspaces)` so I can't tell if this is a `DistributedArrays` bug).

And what I would have thought to be a more natural method:
```
Base.similar(R::JetDSpace{T,S}, dims::NTuple{N,Int}) where {T,S,N} = JetDSpace(similar(R.blkspaces,dims))
```
Errors with:
```
BoundsError: attempt to access 0-element Vector{UnitRange{Int64}} at index [0]
```
because `dims = (0,)`!

So, all in all, something is fishy :)

PS: thanks for a great set of packages, and examples!

**Comment by samtkaplan** [comment]
Hi @raminammour,

Thanks for reporting, and thanks for thinking about a possible solution.  There is, indeed, a missing method.  The specific case that isn't implemented is for when the domain of the operator is distributed block array.  In the context of Jets, the `similar` method is used, here, to create a zero-sized array with a type that is consistent with the domain of the operator.   That array is the linearization point for the jet.  The zero-size is just to save on memory when we don't care about the linearization point (e.g. when the operator, itself, is linear).  Anyways, I'll try to make a fix.  I'm thinking the most appropriate fix may be to create a new function (e.g. `similar_zerosize`) that is used rather than `similar` with the `dims` argument.  That way, it should be easy to make that zero-sized array with the right type even when the domain is a distributed block array, and without abusing the semantics of the `similar` method.  I'll try to whip up a PR.

For your example, a simple work-around might be to define:

```
Base.similar(R::JetDSpace{T}, dims::NTuple{1,Int}) where {T} = JetSpace(T, dims)
```
But, this may break things for nonlinear operators.

**Comment by raminammour** [comment]
In general, abusing `Base.similar` is scary as it is used in generic code (to obtain temporaries, copies...). But it may not matter in this context, as you say.

**Comment by samtkaplan** [comment]
closed by #15.  feel free to re-open if the problem persists.
